src/hpa.c

```diff
@@ -718,6 +726,11 @@ hpa_alloc_batch(tsdn_t *tsdn, pai_t *self, size_t size, size_t nallocs,
 	size_t nsuccess = hpa_alloc_batch_psset(tsdn, shard, size, nallocs,
 	    results);
 
+	malloc_mutex_lock(tsdn, &shard->mtx);
```

Comment:
This ends up acquiring and releasing the shard mutex twice; instead, we should fold the deferred work computation into the other lock acquistion, by passing `deferred_work_generated` as a parameter to `hpa_alloc_batch_psset`.

Reply:
I can do that, but then I'll have to propagate `deferred_work_generated` all the way to `hpa_try_alloc_batch_no_grow` and assign it after `hpa_shard_maybe_do_deferred_work`. It's because there's no single point in `hpa_try_alloc_batch_no_grow` that holds `shard->mtx`, gets called every time and doesn't potentially modify the outcome of `deferred_work_generated` later.

Reply:
I think that's worthwhile; this is a hot mutex, we should be willing to do some gymnastics to avoid redundant acquisitions.

Reply:
Fixed - I put it into `hpa_try_alloc_batch_no_grow`